### ISSUE TITLE
docker run/create 'cannot specify 64-byte hexadecimal strings' sha256: leakage

### ISSUE THREAD

**Issue by WhisperingChaos** [docker/docker#20972]
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------

Output of `docker version`:

```
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 22:37:33 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 22:37:33 2016
 OS/Arch:      linux/amd64
```

Output of `docker info`:

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 6
Server Version: 1.10.2
Storage Driver: vfs
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.13.0-74-generic
Operating System: Alpine Linux v3.3 (containerized)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.704 GiB
Name: 97b6482c3a10
ID: XUPJ:VGFT:2B7I:HOQU:BJQ7:3L4S:SGTL:ZAOO:Q57R:WXHL:BFKY:UWQ2
WARNING: No swap limit support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

Provide additional environment details (AWS, VirtualBox, physical, etc.):

**List the steps to reproduce the issue:**
1.  Create an image or use an existing one.
2.  Obtain the image's Digest ID:
```
docker inspect --format='{{ .Id }}' -- rjmtest
sha256:97ed62d4dc30d1f53b903dac1e90c8fee4a77a30f0099b42f34f67cb5b17018c
```
3.  Attempt to create a container using either docker run/create but remove the ```sha256:``` prefix:
```
docker run 97ed62d4dc30d1f53b903dac1e90c8fee4a77a30f0099b42f34f67cb5b17018c
docker: Invalid repository name (97ed62d4dc30d1f53b903dac1e90c8fee4a77a30f0099b42f34f67cb5b17018c), cannot specify 64-byte hexadecimal strings.
```
4.  Now remove the container using ```docker rmi``` and the same 64-byte hexadecimal part of the Digest ID that failed when attempting to run/create:
```
docker rmi 97ed62d4dc30d1f53b903dac1e90c8fee4a77a30f0099b42f34f67cb5b17018c
Untagged: rjmtest:latest
Deleted: sha256:97ed62d4dc30d1f53b903dac1e90c8fee4a77a30f0099b42f34f67cb5b17018c
Deleted: sha256:884218f6feb1cee2f842cdbbed25bebfbe13d055fbd3254b1be6241f19eb90fb
```
Also, ```docker inspect``` accepts the same 64-byte hexadecimal part of the Digest ID that failed when attempting to run/create.

**Describe the results you received:**
In some cases, the specified Docker operation can use the 64-byte hexadecimal part of the Digest ID to identify an image and successfully complete.  In other cases, the ```sha256:``` prefix must exist in order for the command to accept the complete ID.

**Describe the results you expected:**
* Given ```docker run/create``` accept Digest IDs without a prefix but less than 64-bytes.
* Given commands such as ```docker rmi/inspect`` accept the 64-byte hexadecimal part of the Digest ID without the prefix.

It seems  to me that run/create and any other image command should accept an image Digest ID without its prefix.  However, this ambiguity needs to be resolved by Docker.

**Provide additional info you think is important:**
Although I appreciate the content based IDs, I don't understand why this low level implementation concern has become 'visible'.  From the perspective of a Developer using Docker, I just want a UUID whose scope/boundary of uniqueness is well understood.  I would recommend removing the ```sha256:``` prefix so the concept of content addressability and how it's implemented stays hidden (encapsulated) within the lower, more implementation specific layers of Docker Engine.  Right now, this notion leaks/bubbles to the surface.

**Comment by thaJeztah** [comment]
@aaronlehmann @tonistiigi any reason why we accept the ID *without* `sha256:` for `docker rmi`, but not for `docker run`?

Error message is generated here https://github.com/docker/docker/blob/acd8baccff10642e1d44a342414cd196b81a4605/reference/reference.go#L194

**Comment by aaronlehmann** [comment]
Thanks for the detailed report.

In general, we don't want to allow unprefixed 64-character hexadecimal image references. The concern is that a tag like `3240943c9ea3f72db51bea0a2428e83f3c5fa1312e19af017d026f9bcf70f84b` could cause ambiguity when there is a content-addressable image ID with a matching digest. The safest way to avoid tags like this from being created or used is to reject references like this at parsing time. Admittedly, `rmi` behaves a bit differently here. This is not a big concern because `rmi` can't be used to create an ambiguous tag.

As for "leakage" of the "sha256:" prefix, we tried to strike a balance here. This prefix is part of internal image references, including the paths where they're stored on disk, and we didn't want to completely hide this internal representation from tools that might need it. But at the same time, it's mostly noise to end-users. Thus, docker commands like `docker images` generally hide the prefix, but it's visible in lower-level commands like `docker inspect` that dump raw information, and directly through the API.

Also, as mentioned above, we don't want to allow ambiguity between tags and digests, so `sha256:` is required for full IDs, and because of this places that display a full ID will include the prefix.

The bottom line is that when dealing with full IDs, you should always include the prefix. Since this prefix is already present in output that includes these full IDs, such as the `docker inspect` command you mentioned, all you need to do is treat that output as opaque.

**Comment by WhisperingChaos** [comment]
@aaronlehmann 

Thank you for your rapid and detailed reply discussing the motivation behind the implementation!  Sorry my feedback below wasn't more timely.

>The concern is that a tag like 3240943c9ea3f72db51bea0a2428e83f3c5fa1312e19af017d026f9bcf70f84b could cause ambiguity when there is a content-addressable image ID with a matching digest.

- 64-byte hexadecimal strings have been excluded from the tag name space prior to 1.10.  For example, ```docker build``` generates the following message when provided with a 64 character tag name:  ```Invalid repository name (97ed62d4dc30d1f53b903dac1e90c8fee4a77a30f0099b42f34f67cb5b17018c), cannot specify 64-byte hexadecimal strings``` Therefore,  I'm uncertain why an overlap would occur between a tag name of 64 characters and a content-addressable image ID of the same length.

  What's more concerning, especially from a security perspective, are tag names whose length is < 64 bytes that mimic (absent the 'sha256:' prefix) content-addressable IDs.  In this situation, images named with these truncated hexadecimal formatted tags impersonate the image assigned the content-addressable ID when a docker CLI request uses a truncated form as an argument to the command.  Issue #8069 discusses this security exploit but unfortunately **CVE-2014-5278** hasn't been published and is absent from Docker's own [CVE list](https://www.docker.com/docker-cve-database).

  Perhaps, in response to the the security issue above, specifying the name space for a Docker identifier, in this case ```sha256:```, represents a solution to this security exploit.  IMO, this reasoning represents the most compelling argument to expose the prefix.

>As for "leakage" of the "sha256:" prefix, we tried to strike a balance here. This prefix is part of internal image references, including the paths where they're stored on disk, and we didn't want to completely hide this internal representation from tools that might need it. 

- Tools should be as indifferent as possible to implementation concerns of their components, otherwise, they're susceptible to breaking changes when one or more components alter their implementation.  The Docker bug fix #21002 demonstrates the danger in leaking implementation concerns.  Prior to 1.10 ```docker run``` accepted a 64-byte UUID.  While preparing for release 1.10, which introduced the ```sha256:``` prefix, some part of  ```docker run``` changed to accommodate this leaked concept which broke support for the traditional 64-byte ID.

- Tools that want or must bind to the internal workings of Docker are more fragile and should have to expend more effort to acquire this knowledge then ones that superficially couple to Docker's API.  Therefore Docker's API should discourage Tools from binding to its implementation concerns by preventing exposure of these concerns unless explicitly requested by the Tool through an API setting/query.  The current effortless exposure of ```sha256:```  not only encourages coupling by Tools that should be unaware of content-addressability but breaks existing ones that treat IDs as merely 64-byte hexadecimal strings. 

- Since content-addressable image ID's are a relatively new concept, IMO, only a small set of Tools concern themselves with it, as compared with the ones that simply view it as a unique, 64-byte identifier.  Therefore the decision to leak content-addressability, IMO, favors a minority position and potentially penalizes the majority one.  That said, I do agree that Tools which currently limit their notion of an image ID to a character string of indeterminate length will be unaffected by the additional prefix.

- Finally, when exposing new concepts sometimes the method chosen can mitigate its effects.  For example, although human aesthetics rebel when considering a ```:sha256``` suffix to communicate content- addressability, computer algorithms don't.  Therefore, appending instead of prefixing ```sha256:``` may have naturally allowed most tools to blissfully ignore (truncate) the change while conveying this new feature to Tools reliant on it.
